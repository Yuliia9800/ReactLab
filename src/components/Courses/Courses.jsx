import React, { useState } from 'react';

import { Button } from 'common';
import { CourseCard, SearchBar } from './components';
import { ADD_NEW_COURSE_BUTTON_TEXT } from 'constants/constants';

function Courses({ coursesList, navigateToCreate }) {
	const [search, setSearch] = useState('');

	const courses = coursesList.filter(
		(course) =>
			course.id.toLowerCase().includes(search.toLowerCase()) ||
			course.title.toLowerCase().includes(search.toLowerCase())
	);

	return (
		<div className='flex flex-col justify-between space-y-5 p-5'>
			<div className='flex justify-between'>
				<SearchBar setSearch={setSearch} />
				<Button
					buttonText={ADD_NEW_COURSE_BUTTON_TEXT}
					className='btn-secondary'
					onClick={navigateToCreate}
				/>
			</div>

			{courses.map((courseData) => (
				<CourseCard
					key={courseData.id}
					title={courseData.title}
					description={courseData.description}
					creationDate={courseData.creationDate}
					authors={courseData.authors}
					duration={courseData.duration}
				/>
			))}
		</div>
	);
}

export default Courses;
