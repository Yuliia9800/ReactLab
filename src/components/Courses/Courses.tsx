import React, { useMemo, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import { Button } from 'common';
import { CoursesList } from 'types';
import { mockedCoursesList, ADD_NEW_COURSE_BUTTON_TEXT } from 'constant';
import { CourseCard, SearchBar } from './components';

const Courses: React.FC = () => {
	const navigate = useNavigate();
	const [coursesList] = useState<CoursesList>(mockedCoursesList);
	const [search, setSearch] = useState('');

	const courses = useMemo(
		() =>
			coursesList.filter(
				(course) =>
					course.id.toLowerCase().includes(search.toLowerCase()) ||
					course.title.toLowerCase().includes(search.toLowerCase())
			),
		[coursesList, search]
	);

	return (
		<div className='flex flex-col justify-between space-y-5 p-5'>
			<div className='flex justify-between'>
				<SearchBar setSearch={setSearch} />
				<Button
					className='btn-secondary'
					onClick={() => navigate('/courses/add')}
					buttonText={ADD_NEW_COURSE_BUTTON_TEXT}
				/>
			</div>

			{courses.map(
				({ id, title, description, creationDate, duration, authors }) => (
					<CourseCard
						key={id}
						id={id}
						title={title}
						description={description}
						creationDate={creationDate}
						authors={authors}
						duration={duration}
					/>
				)
			)}
		</div>
	);
};

export default Courses;
