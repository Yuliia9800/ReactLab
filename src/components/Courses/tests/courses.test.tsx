import React from 'react';
import { Route, Routes } from 'react-router-dom';
import { fireEvent, renderWithProviders, screen } from 'testing';
import Courses from '../Courses';

const mockcourses = [
	{
		id: '1',
		title: 'title1',
		description: 'description1',
		creationDate: '10.10.2022',
		duration: '70',
		authors: ['author1,author2'],
	},
	{
		id: '3',
		title: 'title3',
		description: 'description3',
		creationDate: '10.10.2022',
		duration: '70',
		authors: ['2,4,'],
	},
];

test('should match snapshot', () => {
	const state = {
		courses: mockcourses,
	};
	const { container } = renderWithProviders(<Courses />, {
		state,
	});

	expect(container).toMatchSnapshot();
});

test('should display amount of CourseCard equal length of courses array', () => {
	const state = {
		courses: mockcourses,
	};

	renderWithProviders(<Courses />, { state });

	expect(screen.getAllByTestId('course')).toHaveLength(2);
});

test('should display Empty container if courses array length is 0', () => {
	const state = {
		courses: [],
	};

	renderWithProviders(<Courses />, { state });

	expect(screen.queryAllByTestId('course')).toHaveLength(0);
});

test('should be showed after a click on a button "Add new course"', () => {
	const state = {
		courses: [],
		user: {
			role: 'admin',
		},
	};

	renderWithProviders(
		<Routes>
			<Route path='/courses' element={<Courses />} />
			<Route path='/courses/add' element={<div>CourseForm</div>} />
		</Routes>,
		{ state, route: '/courses' }
	);

	const button = screen.getByRole('button', { name: /add new course/i });

	expect(button).toBeInTheDocument();

	fireEvent.click(button);

	expect(screen.getByText('CourseForm')).toBeInTheDocument();
});
