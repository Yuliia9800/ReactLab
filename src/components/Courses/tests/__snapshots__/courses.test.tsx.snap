// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col justify-between space-y-5 p-5"
  >
    <div
      class="flex justify-between"
    >
      <div
        class="flex space-x-3"
      >
        <div
          class="form-control w-full"
        >
          <input
            class="input-primary input input-md input-sm w-72"
            placeholder="Enter course name or id..."
            type="text"
            value=""
          />
        </div>
        <button
          class="btn-sm btn px-5"
          type="button"
        >
          Search
        </button>
      </div>
    </div>
    <div
      class="card flex min-h-[200px] flex-row justify-between gap-10 bg-base-200 p-5 shadow-xl"
      data-testid="course"
    >
      <div
        class="w-4/6 space-y-2"
      >
        <h1
          class="text-2xl font-bold"
        >
          title1
        </h1>
        <p
          class="text-sm"
        >
          description1
        </p>
      </div>
      <div
        class="flex w-2/6 min-w-[300px] flex-col justify-between space-y-2"
      >
        <div
          class="space-y-2"
        >
          <p
            class="truncate text-sm"
          >
            <b>
              Author: 
            </b>
          </p>
          <p
            class="text-sm"
          >
            <b>
              Duration: 
            </b>
            01:10
             hours
          </p>
          <p
            class="text-sm"
          >
            <b>
              Created: 
            </b>
            10.10.2022
          </p>
        </div>
        <div
          class="flex flex-wrap justify-between "
        >
          <button
            class="btn-sm btn px-5"
            type="button"
          >
            Show Course
          </button>
        </div>
      </div>
    </div>
    <div
      class="card flex min-h-[200px] flex-row justify-between gap-10 bg-base-200 p-5 shadow-xl"
      data-testid="course"
    >
      <div
        class="w-4/6 space-y-2"
      >
        <h1
          class="text-2xl font-bold"
        >
          title3
        </h1>
        <p
          class="text-sm"
        >
          description3
        </p>
      </div>
      <div
        class="flex w-2/6 min-w-[300px] flex-col justify-between space-y-2"
      >
        <div
          class="space-y-2"
        >
          <p
            class="truncate text-sm"
          >
            <b>
              Author: 
            </b>
          </p>
          <p
            class="text-sm"
          >
            <b>
              Duration: 
            </b>
            01:10
             hours
          </p>
          <p
            class="text-sm"
          >
            <b>
              Created: 
            </b>
            10.10.2022
          </p>
        </div>
        <div
          class="flex flex-wrap justify-between "
        >
          <button
            class="btn-sm btn px-5"
            type="button"
          >
            Show Course
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
